### PR TITLE
Use new Identity in API router, authentication and authorization

### PR DESCRIPTION
**Pull request by rabbah**
Introduces the foundation for limited scope keys: an activation only key with no CRUD rights. Commits are incremental and self-contained, working up to the point of using the new Identity type to reject requests to CRUD resources with activation keys.

This builds on #1254 (first two commits will be dropped from this pull request).

### CODE REVIEW

File: core/controller/src/main/scala/whisk/core/entitlement/RemoteEntitlement.scala [hunk 1]
```diff
@@ -22,12 +22,14 @@ import scala.concurrent.duration.DurationInt
...
 import scala.util.Try
 
+import Privilege.Privilege
 import akka.actor.ActorSystem
 import spray.client.pipelining.Get
 import spray.client.pipelining.Post
 import spray.client.pipelining.WithTransformerConcatenation
 import spray.client.pipelining.addHeader
 import spray.client.pipelining.sendReceive
+import spray.client.pipelining.sendReceive$default$3
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/entitlement/RemoteEntitlement.scala]: `import spray.client.pipelining._`?

File: core/controller/src/main/scala/whisk/core/entitlement/RemoteEntitlement.scala [hunk 2]
```diff
@@ -37,12 +39,13 @@ import spray.http.StatusCodes.Unauthorized
 import spray.http.Uri
 import spray.httpx.SprayJsonSupport.sprayJsonUnmarshaller
 import spray.httpx.UnsuccessfulResponseException
-import spray.json._
+import spray.json.DefaultJsonProtocol
 import spray.json.DefaultJsonProtocol._
+import spray.json.pimpString
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/entitlement/RemoteEntitlement.scala]: Keep the imports wildcarded?

File: common/scala/src/main/scala/whisk/core/entity/Subject.scala [hunk 3]
```diff
@@ -17,10 +17,12 @@
 package whisk.core.entity
 
 import scala.util.Try
+
+import spray.json.JsString
 import spray.json.JsValue
 import spray.json.RootJsonFormat
-import spray.json.JsString
 import spray.json.deserializationError
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/entity/Subject.scala]: `import spray.json._`

File: common/scala/src/main/scala/whisk/core/entity/AuthKey.scala [hunk 4]
```diff
@@ -16,6 +16,13 @@
 
 package whisk.core.entity
 
+import scala.util.Try
+
+import spray.json.JsString
+import spray.json.JsValue
+import spray.json.RootJsonFormat
+import spray.json.deserializationError
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/entity/AuthKey.scala]: `import spray.json._`

File: core/controller/src/main/scala/whisk/core/controller/Authenticate.scala [hunk 5]
```diff
@@ -69,8 +73,17 @@ trait Authenticate extends Logging {
...
+                    Try {
+                        Identity(s.subject, EntityName(s.subject()), s.authkey, Privilege.ALL)
+                    } match {
+                        case Success(id) => promise.success(Some(id))
+                        case Failure(t) =>
+                            error(this, s"failed to created identity profile: $t")
+                            promise.failure(t)
+                    }
+                case Success(None) =>
+                    promise.success(None)
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/Authenticate.scala]: I believe we can get rid of the Promise, will try to provide a patch proposal later.

File: core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala [hunk 6]
```diff
@@ -214,11 +206,37 @@ protected[core] abstract class EntitlementService(config: WhiskConfig)(
         grant
     }
 
+    // NOTE: explicit grants do not work with package bindings because the current model
+    // for authorization does not allow for a continuation to check that both the binding
+    // and the references package are both either implicitly or explicitly granted; this is
+    // accepted for the time being however because there exists no external mechanism to create
+    // explicit grants
+    protected def check(subject: Subject, right: Privilege, resource: Resource)(
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala]: Note to self: Understand why we need this.

File: core/controller/src/main/scala/whisk/core/controller/Authenticate.scala (hunk 5 above)
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/Authenticate.scala]: Proposed patch.

```
diff --git a/core/controller/src/main/scala/whisk/core/controller/Authenticate.scala b/core/controller/src/main/scala/whisk/core/controller/Authenticate.scala
index 444f049..325ed09 100644
--- a/core/controller/src/main/scala/whisk/core/controller/Authenticate.scala
+++ b/core/controller/src/main/scala/whisk/core/controller/Authenticate.scala
@@ -56,45 +56,33 @@ trait Authenticate extends Logging {
      * authentication directive.
      */
     def validateCredentials(userpass: Option[UserPass])(implicit transid: TransactionId): Future[Option[Identity]] = {
-        val promise = Promise[Option[Identity]]
-        userpass map { pw =>
+        val future = userpass.map { pw =>
             Future {
-                // inside a future should these fail to validate
                 AuthKey(UUID(pw.user), Secret(pw.pass))
             } flatMap { authkey =>
                 info(this, s"authenticate: ${authkey.uuid}")
                 WhiskAuth.get(authStore, authkey.uuid) map { result =>
                     if (authkey == result.authkey) {
                         info(this, s"authentication valid")
-                        Some(result)
+                        Some(Identity(result.subject, EntityName(result.subject()), result.authkey, Privilege.ALL))
                     } else {
                         info(this, s"authentication not valid")
                         None
                     }
                 }
-            } onComplete {
-                case Success(Some(s)) =>
-                    Try {
-                        Identity(s.subject, EntityName(s.subject()), s.authkey, Privilege.ALL)
-                    } match {
-                        case Success(id) => promise.success(Some(id))
-                        case Failure(t) =>
-                            error(this, s"failed to created identity profile: $t")
-                            promise.failure(t)
-                    }
-                case Success(None) =>
-                    promise.success(None)
-                case Failure(_: NoDocumentException | _: IllegalArgumentException) =>
+            } recover {
+                case _: NoDocumentException | _: IllegalArgumentException =>
                     info(this, s"authentication not valid")
-                    promise.success(None)
-                case Failure(t) =>
-                    info(this, s"authentication error: $t")
-                    promise.failure(t)
+                    None
             }
-        } getOrElse {
+        }.getOrElse {
             info(this, s"credentials are malformed")
-            promise.success(None)
+            Future.successful(None)
+        }
+
+        future.onFailure {
+            case t: Throwable => info(this, s"authentication error: $t")
         }
-        promise.future
+        future
     }
 }
```

File: common/scala/src/main/scala/whisk/core/entity/AuthKey.scala (hunk 4 above)
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/AuthKey.scala]: The explicit imports are a function of the IDE's organize import facility (used for example to import new dependencies or remove unused ones). Is there a way to configure it to use wildcards when the number of imports from a package or class exceeds a threshold? Otherwise this will turn into another edit war.

Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/entity/AuthKey.scala]: You can ignore these comments if you want. But I'ld strongly recommend to configure your IDE to leave Wildcard imports alone and always use the wildcard ones, at least for spray.json. It makes later editing of these classes and files a whole lot easier. Won't block the PR though, just a thing i noticed.

Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/AuthKey.scala]: > But I'ld strongly recommend to configure your IDE to leave Wildcard imports alone

We need to automate and standardize such conventions otherwise...

Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/entity/AuthKey.scala]: Agreed, take the comments as a friendly nudge :)

File: common/scala/src/main/scala/whisk/core/entity/Subject.scala [hunk 7]
```diff
@@ -17,15 +17,19 @@
...
+import whisk.core.entitlement.Privilege.Privilege
+import whisk.core.entitlement.Privilege
 
 protected[core] class Subject private (private val subject: String) extends AnyVal {
     protected[core] def apply() = subject
     protected[core] def namespace = EntityPath(subject)
     protected[entity] def toJson = JsString(subject)
+    protected[core] def toIdentity = Identity(this, EntityName(subject), AuthKey(), Privilege.ALL)
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/entity/Subject.scala]: Not sure about this, as it creates a "dummy" `Identity` (for testing?) which might be misleading in usual coding.

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 8]
```diff
@@ -424,15 +425,22 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
...
         // merge package parameters with action (action parameters supersede), then merge in payload
         val args = { env map { _ ++ action.parameters } getOrElse action.parameters } merge payload
-        val message = Message(transid, s"/actions/invoke/${action.namespace}/${action.name}/${action.rev}", user, activationId.make(), args)
+        val message = Message(
+            transid,
+            FullyQualifiedEntityName(action.namespace, action.name, Some(action.rev())),
+            user.subject,
+            user.authkey,
+            activationId.make(),
+            user.subject.namespace,
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: Should this rather be `user.namespace` (which also currently falls back to `subject` being the namespace anyway)? Would be more in line with where we want to be.

File: common/scala/src/main/scala/whisk/core/entity/Subject.scala (hunk 7 above)
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/Subject.scala]: > Not sure about this, as it creates a "dummy" Identity (for testing?) which might be misleading in usual coding.

Negative. I meant to use this in the authentication trait.

File: core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala (hunk 6 above)
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala]: Got it, @rabbah want to rename this to `checkPrivilege` or thelike. Overloading `check` seems like both are equivalent in effects, while they aren't.

**Comment by rabbah** [comment]
pg2/145.

File: core/controller/src/main/scala/whisk/core/controller/Authenticate.scala [hunk 9]
```diff
@@ -64,17 +64,18 @@ trait Authenticate extends Logging {
                         info(this, s"authentication not valid")
                         None
                     }
+                } recover {
+                    case _: NoDocumentException | _: IllegalArgumentException =>
+                        info(this, s"authentication not valid")
+                        None
+                    case t: Throwable =>
+                        error(this, s"authentication error: $t")
+                        None
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/Authenticate.scala]: This is not semantically preserving i believe. In this case you `recover` any error to a `None` whereas before, only `NoDocumentException` and `IllegalArgumentException` have been recovered. Was this change intended?

File: common/scala/src/main/scala/whisk/core/entity/AuthKey.scala (hunk 4 above)
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/AuthKey.scala]: IDE settings updated for future reviewing pleasure.

**Comment by rabbah** [comment]
pg1/458 re-approved.

**Comment by markusthoemmes** [comment]
rebased and merged in 6152dbad30e9df58763c94a06674def61922c5d0
